OSDOCS-3601: Heterogeneous cluster support (#46224)
openshift/openshift-docs: modules/mixed-arch-creating-arm64-bootimage.adoc

//Module included in the following assemblies
//
//post_installation_configuration/cluster-tasks.adoc

:_content-type: PROCEDURE
[id="mixed-arch-creating-arm64-bootimage"]

= Creating an `arm64` bootimage using the Azure image gallery
 
Before you modify the machine set, you must manually generate the `arm64` bootimage. 
 
.Prerequisites

* Installed the Azure CLI (`az`).
* Configured your Azure portal account.
* Created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 

> .Procedure

Author: @aleskandro Hi Alessandro! PTAL at this procedure when you get the chance, thank you!

. Login to your Azure account: 
+
[source,terminal]
----
az login
----
. Create a storage account and upload the `arm64` virtual hard disk (VHD) to your storage account:
+
[source,terminal]
----

> az storage account create -n <storage-account> -g <resource-group> -l westus --sku Standard_LRS 

Author: Also not sure what `resource-group` is used for?

> az storage account create -n <storage-account> -g <resource-group> -l westus --sku Standard_LRS 

Reviewer: i hardcoded westus as an example region. maybe good to templatize that as well

----
+
Create a storage container using the storage account you generated:
+
[source,terminal]
+
----
az storage container create -n <container> --account-name <storage-account>
----
. You must use the OpenShift installer JSON file to extract the `URL` and `arch64` VHD name:

> .. Extract the `URL` field set it to `RHCOS_vhd_origin_url`:

Reviewer 2: The 'and' conjunction is missing between `field` and `set`

----
.. Extract the `aarch64` VHD name and set it to `Blob_name`:
+
[source,terminal]
----
Blob_name=rhcos-$(./openshift-install coreos print-stream-json | jq -r '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".release')-azure.aarch64.vhd
----
. Generate a shared access signature (SAS) token. Use this token to upload the {op-system} VHD to your storage container: 
+
[source,terminal]
----
end=`date -u -d "30 minutes" '+%Y-%m-%dT%H:%MZ'`
sas=`az storage container generate-sas -n <container> --account-name <storage-account> --https-only --permissions dlrw --expiry $end -o tsv`
----
. Copy the {op-system} VHD into the storage container:
+
[source, terminal]
----
az storage blob copy start --account-name $<storage-account> --sas-token "$sas" \
 --source-uri "$RHCOS_vhd_origin_URL" \
 --destination-blob "$Blob_name" --destination-container $<container>
----
+
You can check the status of the of copying process with the following command:
+
[source,terminal]
----
"az storage blob show -c $<container> -n $Blob_name --account-name $<storage-account> | jq .properties.copy"
----
+
.Example output
[source,terminal]
----
 "completionTime": null,

>  "destinationSnapshot": null,

Reviewer 2: This snippet is missing the brackets at the beginning and the end. It is a json object. Not sure it's intended

 
. Create an image gallery using the following command:
+
[source,terminal]
----
az sig create --resource-group <resource-group> --gallery-name <gallery> 
----
Use the image gallery to create an image definition:
+
[source,terminal]
----
az sig image-definition create --resource-group <resource-group> --gallery-name <gallery> --gallery-image-definition rhcos-arm64 --publisher RedHat --offer arm --sku arm64 --os-type linux --architecture Arm64 --hyper-v-generation V2
----
. To get the URL of the VHD and set it to `RHCOS_vhd_url`, run the following command: 
+
[source,terminal]
----
RHCOS_vhd_urlL=$(az storage blob url --account-name $<storage-account> -c $<container> -n "$Blob_name" -o tsv)
----
. Use the `RHCOS_vhd_url`, your storage account, resource group, and image gallery   to create an image version:
+
[source,terminal]
----
az sig image-version create --resource-group $<resource-group> --gallery-name$<gallery> --gallery-image-definition rhcos-arm64 --gallery-image-version 1.0.0 --os-vhd-storage-account $<storage-account> --os-vhd-uri $RHCOS_vhd_url
----
. Your `arm64` imageID for the `resourceID` parameter is now generated: 
+
.Example `resourceID` 
+
[source,terminal]
----

> /resourceGroups/$<resource-group>/providers/Microsoft.Compute/galleries/$<gallery>/images/$<image-name>/versions/$<image-version> 

Author: What does `image-name` refer to here?

Reviewer: ah yes that would refer to the image-definition created in the previous step, i.e rhcos-arm64. would be good to reflect that

//Module included in the following assemblies
//
//post_installation_configuration/cluster-tasks.adoc

:_content-type: PROCEDURE
[id="mixed-arch-creating-arm64-bootimage_{context}""]

= Creating an `arm64` bootimage using the Azure image gallery
 
To configure your heterogeneous cluster, you must create an `arm64` bootimage and add it to your Azure machine set. The following procedure describes how to manually generate an `arm64` bootimage. 
 
.Prerequisites

* Installed the Azure CLI (`az`).
* Created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 

.Procedure
. Login to your Azure account: 
+
[source,terminal]
----
$ az login
----
. Create a storage account and upload the `arm64` virtual hard disk (VHD) to your storage account. The {product-title} installation program creates a resource group, however, the bootimage can also be uploaded to a custom named resource group: 
+
[source,terminal]
----

> $ az storage account create -n ${STORAGE_ACCOUNT_NAME} -g ${RESOURCE_GROUP} -l westus --sku Standard_LRS <1>

Author: I looked into the formating for the names and the `${}` is correct for using the `az` command. 
That was my mistake, we usually use `<>` in other parts of the docs.

----
+
<1> The `westus` object is an example region. 
+
. Create a storage container using the storage account you generated:
+
[source,terminal]
+
----
$ az storage container create -n <container> --account-name <storage-account>
----
. You must use the {product-title} installation program JSON file to extract the URL and `arch64` VHD name:

> .. Extract the `URL` field and set it to `RHCOS_vhd_origin_url` as the file name:

Reviewer: RHCOS_VHD_ORIGIN_URL, all caps because it is an env variable..

+
[source,terminal]
----

> $ RHCOS_vhd_origin_url=$(./openshift-install coreos print-stream-json | jq -r '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".url')

Author: Also is there a way we can use `jsonpath` instead of `jq`, we are trying to avoid using `jq` commands in our docs.

Reviewer 2: I don't think the output option (-o) is available for the installer, is it @Prashanth684 ?

Reviewer: as of today there is not. @kelbrown20 what typically do other docs use?

Author: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output
Near the bottom of the list shows the alternative:
For - `$ oc get clusterversion -o json|jq ".items[0].spec"`
Use insead- `$ oc get clusterversion -o jsonpath='{.items[0].spec}{"\n"}'

Author: It might be different for `oc` commands

Reviewer: the openshift-install command doesn't support -o like Alessandro said. could we still use jq for this?

Author: Looking into. Thank you for checking!

Author: We can keep the `jq` for the TP, but will have to fix come GA for this feature

----
+
.Example output
[source,terminal]
----
{
 "completionTime": null,
 "destinationSnapshot": null,
 "id": "1fd97630-03ca-489a-8c4e-cfe839c9627d",
 "incrementalCopy": null,
 "progress": "17179869696/17179869696",
 "source": "https://rhcos.blob.core.windows.net/imagebucket/rhcos-411.86.202207130959-0-azure.aarch64.vhd",
 "status": "success", <1>
 "statusDescription": null
}
----
+
<1> If the status parameter displays the `success` object, the copying process is complete. 
 
. Create an image gallery using the following command:
+
[source,terminal]
----
$ az sig create --resource-group <resource-group> --gallery-name <gallery> 
----
Use the image gallery to create an image definition. In the following example command, `rhcos-arm64` is the name of the image definition. 
+
[source,terminal]
----
$ az sig image-definition create --resource-group <resource-group> --gallery-name <gallery> --gallery-image-definition rhcos-arm64 --publisher RedHat --offer arm --sku arm64 --os-type linux --architecture Arm64 --hyper-v-generation V2
----
. To get the URL of the VHD and set it to `RHCOS_vhd_url` as the file name, run the following command: 

> +

Reviewer: RHCOS_VHD_URL

[source,terminal]
----
$ RHCOS_VHD_URL=$(az storage blob url --account-name ${STORAGE_ACCOUNT_NAME} -c ${CONTAINER_NAME} -n "${BLOB_NAME}" -o tsv)
----
. Use the `RHCOS_VHD_URL` file, your storage account, resource group, and image gallery   to create an image version:
+
[source,terminal]
----
$ az sig image-version create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME}--gallery-image-definition rhcos-arm64 --gallery-image-version 1.0.0 --os-vhd-storage-account ${STORAGE_ACCOUNT_NAME} --os-vhd-uri ${RHCOS_VHD_URL}
----
. Your `arm64` imageID for the `resourceID` parameter is now generated: 
+
.Example `resourceID` 
+
[source,terminal]
----

> /resourceGroups/${RESOURCE_GROUP}/providers/Microsoft.Compute/galleries/${GALLERY_NAME}/images/rhcos-arm64/versions/1.0.0

Author: I changed this to use `rhcos-arm` for `${IMAGE_NAME}` and `1.0.0` for `${IMAGE_VERSION}` since those are the example names in lines 103 and 115.
Made it the same for the example yaml in next procedure.

Reviewer: honestly that can be templatized as well..but that's up to you.

> /resourceGroups/${RESOURCE_GROUP}/providers/Microsoft.Compute/galleries/${GALLERY_NAME}/images/rhcos-arm64/versions/1.0.0

Reviewer: to get the resource id, i usually go into the UI and click on the "json view" next to the image version and that shows the uri. @aleskandro is there a better way to get it ? is there a way through cli ?

//Module included in the following assemblies
//
//post_installation_configuration/cluster-tasks.adoc

:_content-type: PROCEDURE

> [id="mixed-arch-creating-arm64-bootimage_{context}""]

Reviewer 3: ```suggestion
[id="mixed-arch-creating-arm64-bootimage_{context}"]
```

= Creating an `arm64` boot image using the Azure image gallery
 
To configure your heterogeneous cluster, you must create an `arm64` boot image and add it to your Azure machine set. The following procedure describes how to manually generate an `arm64` boot image. 
 
.Prerequisites

> * Installed the Azure CLI (`az`).

Reviewer 3: ```suggestion
* You installed the Azure CLI (`az`).
```

> * Created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 

Reviewer 3: ```suggestion
* You created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 
```

.Procedure

> . Login to your Azure account: 

Reviewer 3: ISG nit:
```suggestion
. Log in to your Azure account: 
```

+
[source,terminal]
----
$ az login
----
. Create a storage account and upload the `arm64` virtual hard disk (VHD) to your storage account. The {product-title} installation program creates a resource group, however, the boot image can also be uploaded to a custom named resource group: 
+
[source,terminal]
----
$ az storage account create -n ${STORAGE_ACCOUNT_NAME} -g ${RESOURCE_GROUP} -l westus --sku Standard_LRS <1>
----
+
<1> The `westus` object is an example region. 
+
. Create a storage container using the storage account you generated:
+
[source,terminal]
+
----
$ az storage container create -n ${CONTAINER_NAME} --account-name ${STORAGE_ACCOUNT_NAME}
----
. You must use the {product-title} installation program JSON file to extract the URL and `arch64` VHD name:

> .. Extract the `URL` field and set it to `RHCOS_VHD_ORIGIN_URL` as the file name:

Reviewer 3: Doc guidelines recommend including an explicit instruction to "run" or "enter" the command. E.g.
```suggestion
.. Extract the `URL` field and set it to `RHCOS_VHD_ORIGIN_URL` as the file name by running the following command:
```

+
[source,terminal]
----
$ RHCOS_VHD_ORIGIN_URL=$(./openshift-install coreos print-stream-json | jq -r '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".url')
----

> .. Extract the `aarch64` VHD name and set it to `BLOB_NAME` as the file name:

Reviewer 3: Doc guidelines recommend including an explicit instruction to "run" or "enter" the command. E.g.
```suggestion
.. Extract the `aarch64` VHD name and set it to `BLOB_NAME` as the file name by running the following command:
```

+
[source,terminal]
----
$ BLOB_NAME=rhcos-$(./openshift-install coreos print-stream-json | jq -r '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".release')-azure.aarch64.vhd
----

> . Generate a shared access signature (SAS) token. Use this token to upload the {op-system} VHD to your storage container: 

Reviewer 3: Are these two separate commands in this step? If so, the guidelines suggest they be [broken up into two code blocks](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output), one code block per command. E.g.

```
[source,terminal]
----
$ end=`date -u -d "30 minutes" '+%Y-%m-%dT%H:%MZ'`
----
+
[source,terminal]
----
$ sas=`az storage container generate-sas -n ${CONTAINER_NAME} --account-name ${STORAGE_ACCOUNT_NAME} --https-only --permissions dlrw --expiry $end -o tsv`
----
```

Author: Ooo yeah I believe so, will fix that, thanks!

> . Generate a shared access signature (SAS) token. Use this token to upload the {op-system} VHD to your storage container: 

Reviewer 3: Doc guidelines recommend including an explicit instruction to "run" or "enter" the command. E.g.
```suggestion
. Generate a shared access signature (SAS) token. Use this token to upload the {op-system} VHD to your storage container by running the following command: 
```

+

> [source,terminal]

Reviewer 3: Same comment about breaking this up into two separate code blocks

----
$ end=`date -u -d "30 minutes" '+%Y-%m-%dT%H:%MZ'`
$ sas=`az storage container generate-sas -n ${CONTAINER_NAME} --account-name ${STORAGE_ACCOUNT_NAME} --https-only --permissions dlrw --expiry $end -o tsv`
----

> . Copy the {op-system} VHD into the storage container:

Reviewer 3: Doc guidelines recommend including an explicit instruction to "run" or "enter" the command. E.g.
```suggestion
. Copy the {op-system} VHD into the storage container by entering the following command:
```

+
[source, terminal]
----
$ az storage blob copy start --account-name ${STORAGE_ACCOUNT_NAME} --sas-token "$sas" \
 --source-uri "${RHCOS_VHD_ORIGIN_URL}" \
 --destination-blob "${BLOB_NAME}" --destination-container ${CONTAINER_NAME}
----
+
You can check the status of the copying process with the following command:
+
[source,terminal]
----
$ az storage blob show -c ${CONTAINER_NAME} -n ${BLOB_NAME} --account-name ${STORAGE_ACCOUNT_NAME} | jq .properties.copy
----
+
.Example output
[source,terminal]
----
{
 "completionTime": null,
 "destinationSnapshot": null,
 "id": "1fd97630-03ca-489a-8c4e-cfe839c9627d",
 "incrementalCopy": null,
 "progress": "17179869696/17179869696",
 "source": "https://rhcos.blob.core.windows.net/imagebucket/rhcos-411.86.202207130959-0-azure.aarch64.vhd",
 "status": "success", <1>
 "statusDescription": null
}
----
+
<1> If the status parameter displays the `success` object, the copying process is complete. 
 
. Create an image gallery using the following command:
+
[source,terminal]
----
$ az sig create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME}
----

> Use the image gallery to create an image definition. In the following example command, `rhcos-arm64` is the name of the image definition. 

Reviewer 3: Is this a substep of step 7?

Author: It technically can be its own step, but I wanted it as a substep since its based on the previous step

+
[source,terminal]
----
$ az sig image-definition create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME} --gallery-image-definition rhcos-arm64 --publisher RedHat --offer arm --sku arm64 --os-type linux --architecture Arm64 --hyper-v-generation V2
----
. To get the URL of the VHD and set it to `RHCOS_VHD_URL` as the file name, run the following command: 
+
[source,terminal]
----
$ RHCOS_VHD_URL=$(az storage blob url --account-name ${STORAGE_ACCOUNT_NAME} -c ${CONTAINER_NAME} -n "${BLOB_NAME}" -o tsv)
----
. Use the `RHCOS_VHD_URL` file, your storage account, resource group, and image gallery to create an image version. In the following example, `1.0.0` is the image version.
+
[source,terminal]
----
$ az sig image-version create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME} --gallery-image-definition rhcos-arm64 --gallery-image-version 1.0.0 --os-vhd-storage-account ${STORAGE_ACCOUNT_NAME} --os-vhd-uri ${RHCOS_VHD_URL}
----

> . Your `arm64` imageID for the `resourceID` parameter is now generated. You can access your `resourceID` with the following command: 

Reviewer 3: ```suggestion
. Your `arm64` image ID for the `resourceID` parameter is now generated. You can access your `resourceID` with the following command: 
```
This reads like a verification step, is it? The `.Verification` label would be great to use here!

Author: Oh interesting, I'll have to look into that. I don't think its a Verification becuase its more of a "Getting the resourceID" step, and not really a "Verify/Check that your resourceID is generated" step, but I'll check the procedure the SME gave me as well

Reviewer 3: gotcha, no worries if not!

+
[source,terminal]
----

> az sig image-version show -r $GALLERY_NAME -g $RESOURCE_GROUP -i rhcos-arm64 -e 1.0.0

Reviewer 3: ```suggestion
$ az sig image-version show -r $GALLERY_NAME -g $RESOURCE_GROUP -i rhcos-arm64 -e 1.0.0
```

----

> .Example `resourceID` 

Reviewer 3: ```suggestion
.Example output
```
If this isn't an output just ignore 🙂

----
Use the image gallery to create an image definition. In the following example command, `rhcos-arm64` is the name of the image definition. 
+
[source,terminal]
----
$ az sig image-definition create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME} --gallery-image-definition rhcos-arm64 --publisher RedHat --offer arm --sku arm64 --os-type linux --architecture Arm64 --hyper-v-generation V2
----
. To get the URL of the VHD and set it to `RHCOS_VHD_URL` as the file name, run the following command: 
+
[source,terminal]
----
$ RHCOS_VHD_URL=$(az storage blob url --account-name ${STORAGE_ACCOUNT_NAME} -c ${CONTAINER_NAME} -n "${BLOB_NAME}" -o tsv)
----
. Use the `RHCOS_VHD_URL` file, your storage account, resource group, and image gallery to create an image version. In the following example, `1.0.0` is the image version.
+
[source,terminal]
----
$ az sig image-version create --resource-group ${RESOURCE_GROUP} --gallery-name ${GALLERY_NAME} --gallery-image-definition rhcos-arm64 --gallery-image-version 1.0.0 --os-vhd-storage-account ${STORAGE_ACCOUNT_NAME} --os-vhd-uri ${RHCOS_VHD_URL}
----

> . Your `arm64` image ID for the `resourceID` parameter is now generated. You can access your `resourceID` with the following command: 

Reviewer: instead of `Your `arm64` image ID for the `resourceID` parameter is now generated` , maybe `The arm64 bootimage is created. You can access the id of the image with the following command` ?

might want to call out that the `id` field in the json is what is used for `resourceID` in the machineset

Author: Acked, updating

//Module included in the following assemblies
//
//post_installation_configuration/cluster-tasks.adoc

:_content-type: PROCEDURE
[id="mixed-arch-creating-arm64-bootimage"]

= Creating an `arm64` bootimage using the Azure image gallery
 
Before you modify the machine set, you must manually generate the `arm64` bootimage. 
 
.Prerequisites

* Installed the Azure CLI (`az`).
* Installed the OpenShift CLI (`oc`).
* Configured your Azure portal account.

> * Created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 

Reviewer: not necessarily a prerequisite to have oc and a cluster installed to generate an arm bootimage.

Author: Ah, that was my mistake when messing with the modules. Thank ya!

.Procedure
. Login to your Azure account: 
+
[source,terminal]
----
az login
----

> . Create a Microsoft storage accounts and upload the `arm64` virtual hard disk (VHD) (FIXME: To where?)

Author: The doc you sent me just said upload the `arm64` VHD but I wasn't sure on where its suppose to uploaded to? Do you mean download?

Reviewer: upload the vhd to the storage account i meant. sorry the wording wasn't clear

+
[source,terminal]
----
az storage account create -n mystorageaccount -g MyResourceGroup -l westus --sku Standard_LRS 
----
. Create a storage container using the following command: 
+
[source,terminal]
+
----
az storage container create -n mycontainer --account-name mystorageaccount
----
. Download the {op-system} VHD from the `aarch64` link:https://mirror.openshift.com/pub/openshift-v4/arm64/dependencies/rhcos/pre-release/latest-4.11/[image mirror] page.

> . Generate a shared access signature (SAS) token. Use the the token to upload the {op-system} VHD (FIXME: to where?) using the following command: 

Author: Same thing here, not sure what this upload is to?

Reviewer: upload the vhd to the storage account ( the storage container in the storage account actually)

----
. Create an image gallery using the following example command:
+
[source,terminal]
----
az sig create --resource-group MyResourceGroup --gallery-name myGallery
----
. Create an image definition:
+
[source,terminal]
----
az sig image-definition create --resource-group MyResourceGroup --gallery-name myGallery --gallery-image-definition rhcos-arm64 --publisher RedHat --offer arm --sku arm64 --os-type linux --architecture Arm64 --hyper-v-generation V2
----
. Navigate to your Microsoft Azure portal. To find your recourse ID of the storage account, navigate to *Home* -> *Storage accounts* ->, and then *mystorageaccount*. Click *JSON view* and copy the `id` field. 
. To find the URL of the VHD, navigate to *Home* -> *Storage accounts* -> *mystorageaccount* -> *Containers* ->, and then *mycontainer*. Select the VHD you want to upload and copy the `URL` field.
. Use your `ID` and `URL` from your storage account to create the bootimage:
+
[source,terminal]
----
az sig image-version create --resource-group MyResourceGroup --gallery-name myGallery --gallery-image-definition rhcos-arm64 --gallery-image-version 1.0.0 --os-vhd-storage-account <storage-id> --os-vhd-uri <url>
----

>  . (FIXME: I think I need more details on this navigation). Select the version and click JSON View and copy this `ID` field. 

Reviewer: yeah i'm not sure what the best practice here is - i tried to dig into the cli to find an equivalent command that actually reports this information, i'm sure there is but i haven't spent much time on it. Let me try to dig in and find out

//Module included in the following assemblies
//
//post_installation_configuration/cluster-tasks.adoc

:_content-type: PROCEDURE
[id="mixed-arch-creating-arm64-bootimage"]

= Creating an `arm64` bootimage using the Azure image gallery
 
Before you modify the machine set, you must manually generate the `arm64` bootimage. 
 
.Prerequisites

* Installed the Azure CLI (`az`).
* Configured your Azure portal account.
* Created a single-architecture Azure installer-provisioned cluster with the heterogeneous installer binary. 

.Procedure
. Login to your Azure account: 
+
[source,terminal]
----
az login
----
. Create a Microsoft storage account and upload the `arm64` virtual hard disk (VHD) to your resource group:
+
[source,terminal]
----
az storage account create -n <storage-account-name> -g <resource-group-name> -l westus --sku Standard_LRS 
----
. Create a storage container with your storage account using the following command: 
+
[source,terminal]
+
----
az storage container create -n <container-name> --account-name <storage-account-name>
----

> . Download the {op-system} Azure VHD from the `aarch64` link:https://mirror.openshift.com/pub/openshift-v4/arm64/dependencies/rhcos/pre-release/latest-4.11/[image mirror] page.

Author: @Prashanth684 Is there a way to use this image mirror page for the commands that reference RHCOS VHD and the blob name. Its all good If that doesn't work, I was thinking if there was a way to condense the procedure by using the mirror page (If not using the cli is easier here, Id rather do that)
